File: src/fixpm/packages.py
"""npm registry client with graceful offline degradation.

Uses stdlib urllib (zero dependencies). Every network error is swallowed and
degrades to a small curated typo map, so the CLI never crashes without a
connection.
"""

from __future__ import annotations

import json
import math
import urllib.error
import urllib.request
from dataclasses import dataclass
from urllib.parse import quote, urlencode

from . import __version__
from .distance import similarity

SEARCH_URL = "https://registry.npmjs.org/-/v1/search"
DOWNLOAD_URL = "https://api.npmjs.org/downloads/point/last-week/{pkg}"
USER_AGENT = f"fixpm/{__version__} (https://github.com/fixpm/fixpm)"

# Curated high-signal typos so the tool stays useful fully offline.
POPULAR_FALLBACK: dict[str, str] = {
    "loadash": "lodash", "lodass": "lodash", "lodahs": "lodash",
    "expresss": "express", "exprses": "express", "epxress": "express",
    "reactt": "react", "reat": "react", "raect": "react",
    "vuee": "vue", "vu": "vue",
    "angualr": "angular",
    "axioss": "axios", "axois": "axios",
    "momentt": "moment",
    "typsecript": "typescript", "typescirpt": "typescript",
    "eslit": "eslint", "esint": "eslint", "eslintt": "eslint",
    "jestt": "jest",
    "nodemonn": "nodemon",
    "mongose": "mongoose", "mongoos": "mongoose",
    "sequilize": "sequelize",
    "dotnev": "dotenv",
    "tailwindcsss": "tailwindcss",
    "graphl": "graphql",
    "prettie": "prettier",
}

# Local corpus of popular package names. The npm search API has poor recall
# for misspelled queries (searching "loadash" never surfaces "lodash"), so a
# local pool gives transposition-style typos something to fuzzy-match against,
# fully offline. Contributors: append well-known names freely.
POPULAR_PACKAGES: tuple[str, ...] = (
    "react", "react-dom", "react-router", "vue", "vue-router", "pinia",
    "angular", "svelte", "next", "nuxt", "astro", "remix", "gatsby",
    "express", "koa", "fastify", "hapi", "nest", "socket.io", "ws",
    "lodash", "underscore", "ramda", "axios", "superagent", "ky", "got",
    "moment", "dayjs", "date-fns", "luxon",
    "typescript", "eslint", "prettier", "biome", "jest", "vitest", "mocha",
    "chai", "sinon", "playwright", "cypress",
    "vite", "webpack", "rollup", "esbuild", "parcel", "swc", "turbo",
    "tailwindcss", "bootstrap", "bulma", "sass", "less", "postcss",
    "styled-components", "emotion", "framer-motion",
    "three", "d3", "chart.js", "jquery", "gsap",
    "redux", "zustand", "jotai", "immer", "rxjs",
    "zod", "yup", "joi", "ajv", "class-validator",
    "mongoose", "prisma", "sequelize", "typeorm", "knex", "redis",
    "graphql", "trpc", "urql",
    "nodemon", "concurrently", "pm2",
    "dotenv", "cross-env", "commander", "yargs", "inquirer", "prompts",
    "chalk", "kleur", "picocolors", "ora",
    "debug", "winston", "pino", "loglevel",
    "uuid", "nanoid",
    "js-yaml", "yaml", "fs-extra", "glob", "rimraf", "mkdirp",
    "semver", "minimist",
    "jsonwebtoken", "bcrypt", "passport", "helmet", "cors", "multer",
    "body-parser", "compression", "morgan", "cookie-parser", "http-proxy",
)

# Popularity prior for corpus-only hits whose live download count is unknown.
CORPUS_POP_PRIOR = 0.5

@dataclass(frozen=True)
class Suggestion:
    name: str
    weekly_downloads: int  # 0 when unknown
    score: float


def base_name(spec: str) -> str:
    """Strip a trailing version specifier: ``lodash@^4`` -> ``lodash``,
    ``@scope/pkg@1.2`` -> ``@scope/pkg``. A bare scope prefix is kept."""
    head, _, _tail = spec.rpartition("@")
    # head == '' means the only @ was the leading scope marker — nothing to strip
    return head if head else spec


def format_downloads(n: int | None) -> str:
    if n is None:
        return "?"
    if n >= 1_000_000:
        return f"{n / 1_000_000:.1f}M"
    if n >= 1_000:
        return f"{n / 1_000:.0f}k"
    return str(n)


class NpmRegistry:
    """Search-backed suggester. Score = 0.62 * similarity + 0.38 * popularity."""

    def __init__(self, timeout: float = 2.5):
        self.timeout = timeout
        self._cache: dict[str, object] = {}
        self._offline = False  # circuit breaker after repeated failures

    # -- network -----------------------------------------------------------

    def _get_json(self, url: str) -> object | None:
        if self._offline:
            return None
        if url in self._cache:
            return self._cache[url]
        req = urllib.request.Request(
            url, headers={"User-Agent": USER_AGENT, "Accept": "application/json"}
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except (
            urllib.error.URLError,
            TimeoutError,
            OSError,
            json.JSONDecodeError,
            ValueError,
        ):
            if len(self._cache) == 0:
                self._offline = True  # nothing ever succeeded; stop trying
            return None
        if len(self._cache) > 256:
            self._cache.clear()
        self._cache[url] = data
        return data

    def search(self, name: str, size: int = 8) -> list[str]:
        url = f"{SEARCH_URL}?{urlencode({'text': name, 'size': size})}"
        data = self._get_json(url)
        objects = data.get("objects", []) if isinstance(data, dict) else []
        names = []
        for obj in objects:
            pkg = obj.get("package", {})
            if isinstance(pkg.get("name"), str):
                names.append(pkg["name"])
        return names

    def weekly_downloads(self, names: list[str]) -> dict[str, int | None]:
        out: dict[str, int | None] = {}
        for name in names:
            url = DOWNLOAD_URL.format(pkg=quote(name, safe="@"))
            data = self._get_json(url)
            value = data.get("downloads") if isinstance(data, dict) else None
            out[name] = value if isinstance(value, int) else None
        return out

    # -- public API --------------------------------------------------------

    def suggest(self, name: str, limit: int = 3) -> list[Suggestion]:
        target = base_name(name).lower()
        # Curated map wins first: npm search has poor recall for misspelled
        # queries (searching "loadash" never surfaces "lodash"), so classic
        # typos must not depend on the network at all.
        curated = POPULAR_FALLBACK.get(target)
        if curated is not None:
            return [Suggestion(curated, 0, 0.95)]

        registry_names = self.search(target)[:5]
        if any(n.lower() == target for n in registry_names):
            return []  # exact hit: the name was fine, a rename can't be the fix
        downloads = self.weekly_downloads(registry_names)
        # Merge in close corpus matches; they keep the tool useful offline and
        # cover typo shapes the search API cannot recall.
        corpus_hits = [
            p for p in POPULAR_PACKAGES
            if similarity(target, p.lower()) >= 0.5
        ][:3]

        scored: list[Suggestion] = []
        seen: set[str] = set()
        entries = [(n, False) for n in registry_names] \
            + [(p, True) for p in corpus_hits]
        for cand, from_corpus in entries:
            low = cand.lower()
            if low == target or low in seen:
                continue  # exact match means the name was fine; no rename fix
            seen.add(low)
            sim = similarity(target, low)
            if sim < 0.4:
                continue
            dl = downloads.get(cand)
            if dl is not None:
                pop = min(1.0, math.log1p(dl) / math.log1p(3_000_000))
            else:
                dl = 0
                pop = CORPUS_POP_PRIOR if from_corpus else 0.0
            suggestion = Suggestion(cand, dl, round(0.62 * sim + 0.38 * pop, 3))
            if suggestion.score >= 0.4:
                scored.append(suggestion)
        scored.sort(key=lambda s: (-s.score, s.name))
        return scored[:limit]

File: go/packages.go
// npm registry client with graceful offline degradation — port of
// src/fixpm/packages.py.
//
// Network behaviour deliberately differs from the Python original in one way:
// weekly-download lookups are fetched concurrently and bounded by a deadline,
// because fixpm runs from a prompt hook and the Python version issues up to
// five serial HTTPS requests before the prompt can redraw.
package main

import (
	"context"
	"encoding/json"
	"io"
	"math"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// Suggestion mirrors fixpm.packages.Suggestion.
type Suggestion struct {
	Name            string
	WeeklyDownloads int64 // 0 when unknown
	Score           float64
}

// baseName mirrors packages.base_name: strip a trailing version specifier, but
// keep a bare scope prefix ("@scope/pkg" has no version to strip).
func baseName(spec string) string {
	if i := strings.LastIndex(spec, "@"); i > 0 {
		return spec[:i]
	}
	return spec
}

// formatDownloads mirrors packages.format_downloads.
func formatDownloads(n int64) string {
	switch {
	case n >= 1_000_000:
		return formatFloat(float64(n)/1_000_000, 1) + "M"
	case n >= 1_000:
		return formatFloat(float64(n)/1_000, 0) + "k"
	default:
		return itoa(n)
	}
}

// NpmRegistry mirrors packages.NpmRegistry, plus a total time budget.
type NpmRegistry struct {
	timeout time.Duration // per-request ceiling
	budget  time.Duration // whole-run ceiling; 0 means unlimited
	started time.Time

	mu      sync.Mutex
	cache   map[string]any
	offline bool // circuit breaker after repeated failures
}

// NewNpmRegistry mirrors NpmRegistry(timeout=2.5).
func NewNpmRegistry(timeout time.Duration) *NpmRegistry {
	return &NpmRegistry{
		timeout: timeout,
		started: time.Now(),
		cache:   map[string]any{},
	}
}

// WithBudget bounds all registry access to *budget* in total.
//
// This exists because fixpm runs from a prompt hook: the Python original can
// issue up to five serial 2.5s requests before the prompt redraws. A budget
// caps that. Once it is spent, lookups degrade to the offline path instead of
// stalling the shell.
func (r *NpmRegistry) WithBudget(budget time.Duration) *NpmRegistry {
	r.budget = budget
	r.started = time.Now()
	return r
}

// remaining returns the per-request timeout to use, or false when the budget
// is already spent.
func (r *NpmRegistry) remaining() (time.Duration, bool) {
	if r.budget <= 0 {
		return r.timeout, true
	}
	left := r.budget - time.Since(r.started)
	if left <= 0 {
		return 0, false
	}
	if r.timeout > 0 && left > r.timeout {
		left = r.timeout
	}
	return left, true
}

func (r *NpmRegistry) markFailure() {
	r.mu.Lock()
	defer r.mu.Unlock()
	// Nothing ever succeeded -> stop trying for the rest of the process.
	if len(r.cache) == 0 {
		r.offline = true
	}
}

// getJSON mirrors NpmRegistry._get_json.
func (r *NpmRegistry) getJSON(rawURL string) any {
	r.mu.Lock()
	if r.offline {
		r.mu.Unlock()
		return nil
	}
	if v, ok := r.cache[rawURL]; ok {
		r.mu.Unlock()
		return v
	}
	r.mu.Unlock()

	perRequest, ok := r.remaining()
	if !ok {
		return nil // budget spent: degrade instead of stalling the prompt
	}

	ctx, cancel := context.WithTimeout(context.Background(), perRequest)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, rawURL, nil)
	if err != nil {
		r.markFailure()
		return nil
	}
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Accept", "application/json")

	client := &http.Client{Timeout: perRequest}
	resp, err := client.Do(req)
	if err != nil {
		r.markFailure()
		return nil
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode >= 400 {
		// Python's urlopen raises HTTPError here, which _get_json catches.
		r.markFailure()
		return nil
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		r.markFailure()
		return nil
	}
	var data any
	if err := json.Unmarshal(body, &data); err != nil {
		r.markFailure()
		return nil
	}

	r.mu.Lock()
	if len(r.cache) > 256 {
		r.cache = map[string]any{}
	}
	r.cache[rawURL] = data
	r.mu.Unlock()
	return data
}

// search mirrors NpmRegistry.search.
func (r *NpmRegistry) search(name string, size int) []string {
	q := url.Values{}
	q.Set("text", name)
	q.Set("size", itoa(int64(size)))
	rawURL := searchURL + "?" + q.Encode()

	data := r.getJSON(rawURL)
	names := []string{}
	root, ok := data.(map[string]any)
	if !ok {
		return names
	}
	objects, ok := root["objects"].([]any)
	if !ok {
		return names
	}
	for _, item := range objects {
		obj, ok := item.(map[string]any)
		if !ok {
			continue
		}
		pkg, ok := obj["package"].(map[string]any)
		if !ok {
			continue
		}
		if name, ok := pkg["name"].(string); ok {
			names = append(names, name)
		}
	}
	return names
}

// weeklyDownloads mirrors NpmRegistry.weekly_downloads, but issues the lookups
// concurrently so the whole batch costs one round trip instead of N.
func (r *NpmRegistry) weeklyDownloads(names []string) map[string]int64 {
	out := make(map[string]int64, len(names))
	var mu sync.Mutex
	var wg sync.WaitGroup

	for _, name := range names {
		wg.Add(1)
		go func(name string) {
			defer wg.Done()
			rawURL := strings.ReplaceAll(downloadURL, "{pkg}", PyQuote(name, "@"))
			data := r.getJSON(rawURL)

			mu.Lock()
			defer mu.Unlock()
			root, ok := data.(map[string]any)
			if !ok {
				return
			}
			if v, ok := root["downloads"]; ok {
				if f, ok := v.(float64); ok { // encoding/json decodes numbers as float64
					out[name] = int64(f)
				}
			}
		}(name)
	}
	wg.Wait()
	return out
}

// Suggest mirrors NpmRegistry.suggest.
func (r *NpmRegistry) Suggest(name string, limit int) []Suggestion {
	target := strings.ToLower(baseName(name))

	// Curated map wins first: npm search has poor recall for misspelled
	// queries, so classic typos must not depend on the network at all.
	if curated, ok := popularFallback[target]; ok {
		return []Suggestion{{Name: curated, WeeklyDownloads: 0, Score: 0.95}}
	}

	registryNames := r.search(target, 8)
	if len(registryNames) > 5 {
		registryNames = registryNames[:5]
	}
	for _, n := range registryNames {
		if strings.ToLower(n) == target {
			return nil // exact hit: the name was fine, a rename can't be the fix
		}
	}

	downloads := r.weeklyDownloads(registryNames)

	corpusHits := []string{}
	for _, p := range popularPackages {
		if Similarity(target, strings.ToLower(p)) >= 0.5 {
			corpusHits = append(corpusHits, p)
			if len(corpusHits) == 3 {
				break
			}
		}
	}

	type entry struct {
		name       string
		fromCorpus bool
	}
	entries := make([]entry, 0, len(registryNames)+len(corpusHits))
	for _, n := range registryNames {
		entries = append(entries, entry{n, false})
	}
	for _, p := range corpusHits {
		entries = append(entries, entry{p, true})
	}

	logScale := math.Log1p(3_000_000)
	scored := []Suggestion{}
	seen := map[string]bool{}
	for _, e := range entries {
		low := strings.ToLower(e.name)
		if low == target || seen[low] {
			continue // exact match means the name was fine; no rename fix
		}
		seen[low] = true

		sim := Similarity(target, low)
		if sim < 0.4 {
			continue
		}

		dl, hasDL := downloads[e.name]
		var pop float64
		if hasDL {
			pop = math.Min(1.0, math.Log1p(float64(dl))/logScale)
		} else {
			dl = 0
			if e.fromCorpus {
				pop = corpusPopPrior
			}
		}

		s := Suggestion{
			Name:            e.name,
			WeeklyDownloads: dl,
			Score:           PyRound(0.8*sim+0.2*pop, 3),
		}
		if s.Score >= 0.4 {
			scored = append(scored, s)
		}
	}

	sortSuggestions(scored)
	if len(scored) > limit {
		scored = scored[:limit]
	}
	return scored
}
